### ISSUE TITLE
NodePort not working with custom network configs

### ISSUE THREAD

**Issue by benabineri** [aws/amazon-vpc-cni-k8s#593]
Hi,

We are having issues with NodePorts on EKS.

If a node is running the appropriate pod, e.g. an nginx ingress controller, then the nodeport on that node works. Nodeports on nodes not running the pod don't work.

We are running the following versions of the various components:

```
aws-node-5nt5s                                       602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.5.3
aws-node-df9g6                                       602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.5.3
aws-node-fc8vd                                       602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.5.3
aws-node-tmlss                                       602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:v1.5.3
calico-node-f66f4                                    quay.io/calico/node:v3.8.1
calico-node-hk4n6                                    quay.io/calico/node:v3.8.1
calico-node-rfshz                                    quay.io/calico/node:v3.8.1
calico-node-tt7m8                                    quay.io/calico/node:v3.8.1
calico-typha-55cffb48d9-qpz6h                        quay.io/calico/typha:v3.8.1
calico-typha-horizontal-autoscaler-7f77789b9f-w946x  k8s.gcr.io/cluster-proportional-autoscaler-amd64:1.1.2
kube-proxy-5cz2r                                     602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-proxy:v1.13.7
kube-proxy-5jfn6                                     602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-proxy:v1.13.7
kube-proxy-77tbv                                     602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-proxy:v1.13.7
kube-proxy-q85nc                                     602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-proxy:v1.13.7
```

We are using custom network configs and external SNAT. The CNI is configured as follows:
```
      AWS_VPC_CNI_NODE_PORT_SUPPORT:       true
      AWS_VPC_K8S_CNI_LOGLEVEL:            DEBUG
      AWS_VPC_K8S_CNI_EXTERNALSNAT:        true
      AWS_VPC_K8S_CNI_CUSTOM_NETWORK_CFG:  true
      ENI_CONFIG_LABEL_DEF:                failure-domain.beta.kubernetes.io/zone
      MY_NODE_NAME:                         (v1:spec.nodeName)
      WATCH_NAMESPACE:                     kube-system (v1:metadata.namespace)
```

Some iptables digging on a node through which the service _isn't_ reachable has shown the following:

```
# grep ID=3488 /var/log/messages | cut -d' ' -f 7
raw:PREROUTING:rule:2
raw:cali-PREROUTING:rule:1
raw:cali-PREROUTING:rule:3
raw:cali-from-host-endpoint:return:1
raw:cali-PREROUTING:return:5
raw:PREROUTING:policy:3
mangle:PREROUTING:rule:1
mangle:cali-PREROUTING:rule:3
mangle:cali-from-host-endpoint:return:1
mangle:cali-PREROUTING:return:5
mangle:PREROUTING:rule:2
mangle:PREROUTING:policy:4
nat:PREROUTING:rule:1
nat:cali-PREROUTING:rule:1
nat:cali-fip-dnat:return:1
nat:cali-PREROUTING:return:2
nat:PREROUTING:rule:2
nat:KUBE-SERVICES:rule:119
nat:KUBE-NODEPORTS:rule:6
nat:KUBE-XLB-QDXGRUFK7NEGZ3EP:rule:2
nat:KUBE-MARK-DROP:rule:1
nat:KUBE-MARK-DROP:return:2
nat:KUBE-XLB-QDXGRUFK7NEGZ3EP:return:3
nat:KUBE-NODEPORTS:return:17
nat:KUBE-SERVICES:return:120
nat:PREROUTING:policy:3
mangle:INPUT:policy:1
filter:INPUT:rule:1
filter:cali-INPUT:rule:3
filter:cali-INPUT:rule:4
filter:cali-from-host-endpoint:return:1
filter:cali-INPUT:return:6
filter:INPUT:rule:2
filter:KUBE-SERVICES:return:12
filter:INPUT:rule:3
filter:KUBE-EXTERNAL-SERVICES:return:2
filter:INPUT:rule:4
filter:KUBE-FIREWALL:rule:1
```

My understanding is that KUBE-MARK-DROP marks the packet and then KUBE-FIREWALL drops it, but in the intervening period something related to node ports is supposed to unmark it and cause it to be forwarded. I may be misunderstanding as I am not an iptables expert...

I have a development cluster which I can experiment freely on so if there is any more information that would help or you would like me try to anything, please just ask.

Thanks

**Comment by mogren** [comment]
Hey @benabineri, thanks for the well written question. Many of these rules are set up by Calico, and I'm afraid I don't know enough about that yet.

Are you saying that just having Services with [Type=NodePort](https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-nodeport) don't work, but if you have at least one Service with [Type=LoadBalancer](https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-loadbalancer) on the node it all works?

I tried looking for related questions in the Calico issue tracker, but couldn't find anything, so my guess is that it's something in the service configuration that is causing this.

**Comment by benabineri** [comment]
Thanks for your response.

As happens with these things, after I opened this issue I took a step back from iptables and spotted that at some point all of the NodePort services had been set to `externalTrafficPolicy: Local`, so dropped traffic is exactly what we were asking for.

I set the appropriate services to `Cluster` and they're now behaving as expected. 

I'm not sure how it had been working before, we were running a custom CNI image (related to https://github.com/aws/amazon-vpc-cni-k8s/pull/340) but with cross-AZ load balancing on the NLB *off*, so I'm not sure how traffic was being routed between nodes.

I think we're going to go with `externalTrafficPolicy: Cluster` and cross-AZ load balancing on the NLB, for improved redundancy at the cost of routing overhead.
